Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs

```diff
@@ -891,129 +893,58 @@ public void DisableAllServicesByTypeAndName(Type interfaceType, string serviceNa
-            foreach (var service in registeredMixedRealityServices)
-            {
-                service.Item2.Reset();
-            }
+            ExecuteOnAllServices(service => service.Reset());
         }
 
         private void EnableAllServices()
         {
-            // If the Mixed Reality Toolkit is not configured, stop.
-            if (activeProfile == null) { return; }
-
-            // If the Mixed Reality Toolkit is not initialized, stop.
-            if (!IsInitialized) { return; }
-
             // Enable all systems
-            foreach (var system in activeSystems)
-            {
-                system.Value.Enable();
-            }
-
-            // Reset all registered runtime services
-            foreach (var service in registeredMixedRealityServices)
-            {
-                service.Item2.Enable();
-            }
+            ExecuteOnAllServices(service => service.Enable());
         }
 
         private void UpdateAllServices()
         {
-            // If the Mixed Reality Toolkit is not configured, stop.
-            if (activeProfile == null) { return; }
-
-            // If the Mixed Reality Toolkit is not initialized, stop.
-            if (!IsInitialized) { return; }
-
             // Update all systems
-            foreach (var system in activeSystems)
-            {
-                system.Value.Update();
-            }
-
-            // Update all registered runtime services
-            foreach (var service in registeredMixedRealityServices)
-            {
-                service.Item2.Update();
-            }
+            ExecuteOnAllServices(service => service.Update());
         }
 
         private void DisableAllServices()
         {
-            // If the Mixed Reality Toolkit is not configured, stop.
-            if (activeProfile == null) { return; }
-
-            // If the Mixed Reality Toolkit is not initialized, stop.
-            if (!IsInitialized) { return; }
-
             // Disable all systems
-            foreach (var system in activeSystems)
-            {
-                system.Value.Disable();
-            }
-
-            // Disable all registered runtime services
-            foreach (var service in registeredMixedRealityServices)
-            {
-                service.Item2.Disable();
-            }
+            ExecuteOnAllServices(service => service.Disable());
         }
 
         private void DestroyAllServices()
         {
-            // If the Mixed Reality Toolkit is not configured, stop.
-            if (activeProfile == null) { return; }
+            if (!ExecuteOnAllServices(service => service.Destroy())) { return; }
 
-            // If the Mixed Reality Toolkit is not initialized, stop.
-            if (!IsInitialized) { return; }
+            activeSystems.Clear();
+            registeredMixedRealityServices.Clear();
+        }
+
+        private bool ExecuteOnAllServices(Action<IMixedRealityService> execute)
```

Comment:
What was the decision to use an action instead of just calling the methods directly?

This would cut down on the runtime overhead and allocations. Good to remember that much of this code will run on mobile devices.

Have we done a perf comparison?

Reply:
Agreed, due to the tight performance marks on the operations of the services, Actions should be avoided unless absolutely necessary

Reply:
Are you sure you're not over-optimizing this at the expense of readability and maintainability? Those lambdas are executed not even 20 times per frame.
Just found this Stackoverflow thread which tested performance differences between various possible ways of calling methods and the numbers are so tiny, I don't think they really matter.

https://stackoverflow.com/questions/35805609/performance-of-expression-compile-vs-lambda-direct-vs-virtual-calls

If it does matter, I'd like to know how you guys test code performance as this would be critical to know for everyone who wants to participate in improving MRTK.

Reply:
Agree perf is critical in services but I would need to see a stopwatch test showing that this approach actually has a measurable impact before ruling it out.

Reply:
Was wondering if this receives a yes or nor while I was cleaning up my branches

Reply:
@Alexees, the team was very busy preparing the RC1 release and this PR was postponed.

I will personally pull and test your changes this week.

Thanks!